src/vm/amd64/pinvokestubs.S

```diff
@@ -0,0 +1,277 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information. 
+//
+
+.intel_syntax noprefix
+#include "unixasmmacros.inc"
+#include "asmconstants.h"
+
+PInvokeStubForHost_CALLEE_SCRATCH_SIZE = 20h
+
+PInvokeStubForHost_STACK_FRAME_SIZE = PInvokeStubForHost_CALLEE_SCRATCH_SIZE
+
+// 8 FP parameter registers
+PInvokeStubForHost_XMM_SAVE_OFFSET = PInvokeStubForHost_STACK_FRAME_SIZE
+PInvokeStubForHost_STACK_FRAME_SIZE = PInvokeStubForHost_STACK_FRAME_SIZE + 80h
+
+// Ensure that the new rsp will be 16-byte aligned.
+.if ((PInvokeStubForHost_STACK_FRAME_SIZE + 8) % 16) != 0
+PInvokeStubForHost_STACK_FRAME_SIZE = PInvokeStubForHost_STACK_FRAME_SIZE + 8
+.endif
+
+// Return address is immediately above the local variables.
+PInvokeStubForHost_RETURN_ADDRESS_OFFSET = PInvokeStubForHost_STACK_FRAME_SIZE
+PInvokeStubForHost_PARAM_REGISTERS_OFFSET = PInvokeStubForHost_RETURN_ADDRESS_OFFSET + 8
+
+NESTED_ENTRY PInvokeStubForHost, _TEXT, NoHandler
+        alloc_stack     PInvokeStubForHost_STACK_FRAME_SIZE
+        END_PROLOGUE
+
+        // spill args
+        mov             [rsp + PInvokeStubForHost_PARAM_REGISTERS_OFFSET +  0h], rdi
+        mov             [rsp + PInvokeStubForHost_PARAM_REGISTERS_OFFSET +  8h], rdx
+        mov             [rsp + PInvokeStubForHost_PARAM_REGISTERS_OFFSET + 10h], rdx
+        mov             [rsp + PInvokeStubForHost_PARAM_REGISTERS_OFFSET + 18h], rcx
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET +  0h], xmm0
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 10h], xmm1
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 20h], xmm2
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 30h], xmm3
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 40h], xmm4
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 50h], xmm5
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 60h], xmm6
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 70h], xmm7
+        
+        
+        // PInvokeStubForHostWorker(#stack args, stack frame, this)
+        mov             rdx, rdi
+        mov             rdi, r11
+        mov             rsi, rsp
+        call            C_FUNC(PInvokeStubForHostWorker)
+        
+        // unspill return value
+        mov             rax,  [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET +  0h]
+        movdqa          xmm0, [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 10h]
```

Comment:
Please add a UNIXTODO: here to add unspilling rdx and xmm1 when returning structs in registers on Unix is implemented.